# Trabajo Práctico 1 - Laboratorio de Datos - 1C 2025

**Integrantes del grupo:** Julian Hardie, Walter Yance, Rodrigo Gustavo Coppa

In [5]:
# Descomentar y correr esto una sola vez
!pip install plotly --upgrade
!pip install geopandas

   ---------------------------------------- 0.0/14.8 MB ? eta -:--:--
   -- ------------------------------------- 0.8/14.8 MB 4.8 MB/s eta 0:00:03
   ----- ---------------------------------- 2.1/14.8 MB 5.6 MB/s eta 0:00:03
   ------- -------------------------------- 2.9/14.8 MB 4.9 MB/s eta 0:00:03
   -------- ------------------------------- 3.1/14.8 MB 5.0 MB/s eta 0:00:03
   --------- ------------------------------ 3.4/14.8 MB 3.3 MB/s eta 0:00:04
   ------------ --------------------------- 4.5/14.8 MB 3.5 MB/s eta 0:00:03
   ------------- -------------------------- 5.0/14.8 MB 3.6 MB/s eta 0:00:03
   --------------- ------------------------ 5.8/14.8 MB 3.5 MB/s eta 0:00:03
   ------------------ --------------------- 6.8/14.8 MB 3.6 MB/s eta 0:00:03
   ------------------- -------------------- 7.1/14.8 MB 3.5 MB/s eta 0:00:03
   -------------------- ------------------- 7.6/14.8 MB 3.4 MB/s eta 0:00:03
   ---------------------- ----------------- 8.4/14.8 MB 3.3 MB/s eta 0:00:02
   ---

In [8]:
# Si requieren mas librerias, pueden importarlas aca
import numpy as np
import pandas as pd
import geopandas as gpd
import seaborn as sns
import seaborn.objects as so
import json
import plotly.express as px
from formulaic import Formula
from sklearn import linear_model  
from sklearn.metrics import mean_squared_error, r2_score, root_mean_squared_error  

# Si no existe root_mean_squared_error, quitarlo o actualizar scikit-learn con el comando
#!pip install scikit-learn --upgrade

In [14]:
df_educacion = pd.read_csv(r"C:\Users\Usuario\Downloads\indicadores-proceso-trayectoria-sobreedad-2012_2022.csv")
df_educacion

anio  municipio_id municipio_nombre  promocion_efectiva_primaria  \
0     2012          6007    Adolfo Alsina                       100.26   
1     2013          6007    Adolfo Alsina                        97.35   
2     2014          6007    Adolfo Alsina                        97.36   
3     2015          6007    Adolfo Alsina                        96.00   
4     2016          6007    Adolfo Alsina                        97.67   
...    ...           ...              ...                          ...   
1480  2018          6882           Zarate                        95.42   
1481  2019          6882           Zarate                        95.62   
1482  2020          6882           Zarate                        99.51   
1483  2021          6882           Zarate                        96.77   
1484  2022          6882           Zarate                        97.49   

      promocion_efectiva_secundaria  repitencia_primaria  \
0                             84.14                 1.25   
1                             86.84                 1.68   
2                             80.79                 2.23   
3                             91.55                 2.41   
4                             95.56                 2.61   
...                             ...                  ...   
1480                          87.30                 3.83   
1481                          86.58                 3.55   
1482                          99.06                 0.04   
1483                          90.87                 2.79   
1484                          88.91                 1.35   

      repitencia_secundaria  reinscripcion_primaria  reinscripcion_secundaria  \
0                      9.06                    0.07                      1.79   
1                      5.31                    0.13                      0.12   
2                      2.52                    0.07                      0.43   
3                      5.24                    0.21                      0.85   
4                      7.27                    0.07                      1.29   
...                     ...                     ...                       ...   
1480                   8.27                    0.06                      1.80   
1481                   9.53                    0.01                      0.83   
1482                   0.00                    0.00                      0.09   
1483                   6.60                    0.17                      0.64   
1484                   7.13                    0.22                      1.11   

      abandono_interanual_primaria  ...  alumnos_no_promovidos_primaria  \
0                            -1.57  ...                            1.45   
1                             0.84  ...                            0.70   
2                             0.34  ...                            0.69   
3                             1.38  ...                            1.47   
4                            -0.35  ...                            1.58   
...                            ...  ...                             ...   
1480                          0.69  ...                            2.28   
1481                          0.82  ...                            1.97   
1482                          0.45  ...                            0.00   
1483                          0.28  ...                            0.21   
1484                          0.95  ...                            0.99   

      alumnos_no_promovidos_secundaria  salidos_sin_pase_primaria  \
0                                18.54                       0.00   
1                                 9.48                       0.30   
2                                 9.13                       0.13   
3                                 8.62                       0.33   
4                                 6.95                       0.14   
...                                ...                        ...   
1480                             12.39 

In [22]:
df_educacion.head()

anio  municipio_id municipio_nombre  promocion_efectiva_primaria  \
0  2012          6007    Adolfo Alsina                       100.26   
1  2013          6007    Adolfo Alsina                        97.35   
2  2014          6007    Adolfo Alsina                        97.36   
3  2015          6007    Adolfo Alsina                        96.00   
4  2016          6007    Adolfo Alsina                        97.67   

   promocion_efectiva_secundaria  repitencia_primaria  repitencia_secundaria  \
0                          84.14                 1.25                   9.06   
1                          86.84                 1.68                   5.31   
2                          80.79                 2.23                   2.52   
3                          91.55                 2.41                   5.24   
4                          95.56                 2.61                   7.27   

   reinscripcion_primaria  reinscripcion_secundaria  \
0                    0.07                      1.79   
1                    0.13                      0.12   
2                    0.07                      0.43   
3                    0.21                      0.85   
4                    0.07                      1.29   

   abandono_interanual_primaria  ...  alumnos_no_promovidos_primaria  \
0                         -1.57  ...                            1.45   
1                          0.84  ...                            0.70   
2                          0.34  ...                            0.69   
3                          1.38  ...                            1.47   
4                         -0.35  ...                            1.58   

   alumnos_no_promovidos_secundaria  salidos_sin_pase_primaria  \
0                             18.54                       0.00   
1                              9.48                       0.30   
2                              9.13                       0.13   
3                              8.62                       0.33   
4                              6.95                       0.14   

   salidos_sin_pase_secundaria  salidos_con_pase_primaria  \
0                         2.71                       5.40   
1                         2.77                       5.02   
2                         2.11                       4.84   
3                         2.49                       4.96   
4                         2.79                       4.72   

   salidos_con_pase_secundaria  sobreedad_primaria  sobreedad_secundaria  \
0                         2.90               12.61                 26.77   
1                         5.72               10.27                 26.27   
2                         4.31                8.46                 25.06   
3                         4.64                7.16                 30.85   
4                         3.57                6.81                 26.13   

   sobreedad_avanzada_primaria  sobreedad_avanzada_secundaria  
0                         4.01                          12.28  
1                         3.42                          11.51  
2                         2.51                           9.17  
3                         1.86                          10.51  
4                         1.69                           8.24  

[5 rows x 23 columns]

In [24]:
df_educacion.columns

Index(['anio', 'municipio_id', 'municipio_nombre',
       'promocion_efectiva_primaria', 'promocion_efectiva_secundaria',
       'repitencia_primaria', 'repitencia_secundaria',
       'reinscripcion_primaria', 'reinscripcion_secundaria',
       'abandono_interanual_primaria', 'abandono_interanual_secundaria',
       'alumnos_promovidos_primaria', 'alumnos_promovidos_secundaria',
       'alumnos_no_promovidos_primaria', 'alumnos_no_promovidos_secundaria',
       'salidos_sin_pase_primaria', 'salidos_sin_pase_secundaria',
       'salidos_con_pase_primaria', 'salidos_con_pase_secundaria',
       'sobreedad_primaria', 'sobreedad_secundaria',
       'sobreedad_avanzada_primaria', 'sobreedad_avanzada_secundaria'],
      dtype='object')

### Procesamiento de Datos

In [71]:
df_educacion.dtypes

anio                                  int64
municipio_id                          int64
municipio_nombre                     object
promocion_efectiva_primaria         float64
promocion_efectiva_secundaria       float64
repitencia_primaria                 float64
repitencia_secundaria               float64
reinscripcion_primaria              float64
reinscripcion_secundaria            float64
abandono_interanual_primaria        float64
abandono_interanual_secundaria      float64
alumnos_promovidos_primaria         float64
alumnos_promovidos_secundaria       float64
alumnos_no_promovidos_primaria      float64
alumnos_no_promovidos_secundaria    float64
salidos_sin_pase_primaria           float64
salidos_sin_pase_secundaria         float64
salidos_con_pase_primaria           float64
salidos_con_pase_secundaria         float64
sobreedad_primaria                  float64
sobreedad_secundaria                float64
sobreedad_avanzada_primaria         float64
sobreedad_avanzada_secundaria   

**1.** Describir el tipo de variable de cada columna

El tipo de variable de las columnas, en su mayoría, es Numérica y Continua ya que toma valores intermedios como la de **repitencia_primaria** o **abandono_interanual_primaria**, entre otras columnas. Estas representan el porcentaje de alumnos que, por ejemplo, en repitencia_primaria repiten el año para ese nivel educativo en un determinado municipio y un año determinado. 

Para las columnas **anio** y **municipio_id** las variables son tambien numéricas pero discretas ya que no representan valores intermedios. Además, el año tiene un orden específico ya que van del 2012 al 2022 y se vuelve a repetir el ciclo para otro Municipio. También podemos notar que el id del Municipio todas arrancan en 06 (Adolfo Alsina en 06007 es el primero) y van aumentando hasta el ultimo Municipio, que es Zárate, hasta 06882. 

Por último, para la columna **municipio_nombre** la variable es Categórica y Nominal ya que aunque estén ordenadas alfabéticamente, no tienen un orden en si entre ellas. 

**2.** ¿La base de datos contiene datos faltantes? ¿Cuántos?

Se puede ver en la siguiente tabla que no hay filas con datos faltantes. Es decir ninguna de ellas contiene valores nulos o NaN

In [68]:
faltantes = df_educacion.isnull().sum()
faltantes

anio                                0
municipio_id                        0
municipio_nombre                    0
promocion_efectiva_primaria         0
promocion_efectiva_secundaria       0
repitencia_primaria                 0
repitencia_secundaria               0
reinscripcion_primaria              0
reinscripcion_secundaria            0
abandono_interanual_primaria        0
abandono_interanual_secundaria      0
alumnos_promovidos_primaria         0
alumnos_promovidos_secundaria       0
alumnos_no_promovidos_primaria      0
alumnos_no_promovidos_secundaria    0
salidos_sin_pase_primaria           0
salidos_sin_pase_secundaria         0
salidos_con_pase_primaria           0
salidos_con_pase_secundaria         0
sobreedad_primaria                  0
sobreedad_secundaria                0
sobreedad_avanzada_primaria         0
sobreedad_avanzada_secundaria       0
dtype: int64

**3.** Además de la división por municipio, nos interesa saber a que zona de la provincia pertenece cada municipio (Zona Norte, Zona Oeste , Zona Sur e Interior). Para esto:

**(a)** Implementar una función **determinar_zona** que, dado el nombre de un municipio, devuelva a que zona pertencece: ’Norte’, ’Oeste’, ’Sur’ o ’Interior’.

In [79]:
zona_norte = ['San Isidro', 'Vicente Lopez', 'General San Martin', 'San Fernando', 'Tigre']
zona_oeste = ['Moron', 'Tres de Febrero', 'La Matanza', 'Merlo', 'Moreno']
zona_sur = ['Almirante Brown', 'Avellaneda', 'Quilmes', 'Lanus', 'Florencio Varela', 'Berazategui', 'Lomas de Zamora', 'Esteban Echeverria']

In [85]:
def determinar_zona(municipio):
    if municipio in zona_norte:
        return "Norte"
    elif municipio in zona_oeste:
        return "Oeste"
    elif municipio in zona_sur:
        return "Sur"
    else:
        return "Interior"

In [87]:
determinar_zona("San Isidro")

'Norte'

**(b)** Agregar la columna zona a df_educacion que contenga la zona a la que pertenece el municipio.

In [96]:
df_educacion["zona"] = df_educacion["municipio_nombre"].apply(determinar_zona)
df_educacion

anio  municipio_id municipio_nombre  promocion_efectiva_primaria  \
0     2012          6007    Adolfo Alsina                       100.26   
1     2013          6007    Adolfo Alsina                        97.35   
2     2014          6007    Adolfo Alsina                        97.36   
3     2015          6007    Adolfo Alsina                        96.00   
4     2016          6007    Adolfo Alsina                        97.67   
...    ...           ...              ...                          ...   
1480  2018          6882           Zarate                        95.42   
1481  2019          6882           Zarate                        95.62   
1482  2020          6882           Zarate                        99.51   
1483  2021          6882           Zarate                        96.77   
1484  2022          6882           Zarate                        97.49   

      promocion_efectiva_secundaria  repitencia_primaria  \
0                             84.14                 1.25   
1                             86.84                 1.68   
2                             80.79                 2.23   
3                             91.55                 2.41   
4                             95.56                 2.61   
...                             ...                  ...   
1480                          87.30                 3.83   
1481                          86.58                 3.55   
1482                          99.06                 0.04   
1483                          90.87                 2.79   
1484                          88.91                 1.35   

      repitencia_secundaria  reinscripcion_primaria  reinscripcion_secundaria  \
0                      9.06                    0.07                      1.79   
1                      5.31                    0.13                      0.12   
2                      2.52                    0.07                      0.43   
3                      5.24                    0.21                      0.85   
4                      7.27                    0.07                      1.29   
...                     ...                     ...                       ...   
1480                   8.27                    0.06                      1.80   
1481                   9.53                    0.01                      0.83   
1482                   0.00                    0.00                      0.09   
1483                   6.60                    0.17                      0.64   
1484                   7.13                    0.22                      1.11   

      abandono_interanual_primaria  ...  alumnos_no_promovidos_secundaria  \
0                            -1.57  ...                             18.54   
1                             0.84  ...                              9.48   
2                             0.34  ...                              9.13   
3                             1.38  ...                              8.62   
4                            -0.35  ...                              6.95   
...                            ...  ...                               ...   
1480                          0.69  ...                             12.39   
1481                          0.82  ...                             13.63   
1482                          0.45  ...                              0.58   
1483                          0.28  ...                              8.23   
1484                          0.95  ...                             10.75   

      salidos_sin_pase_primaria  salidos_sin_pase_secundaria  \
0                          0.00                         2.71   
1                          0.30                         2.77   
2                          0.13                         2.11   
3                          0.33                         2.49   
4                          0.14                         2.79   
...                         ...                          ...   
1480                       0.09                   

**4.** Quisiéramos analizar si en general la tasa de repitencia es más alta en primaria que en secundaria. Para esto:

**(a)** Agregar una columna **mayor_repitencia_primaria** a df_educacion que indique si ocurre esta situación.

In [135]:
df_educacion["mayor_repitencia_primaria"] = df_educacion["repitencia_primaria"] > df_educacion["repitencia_secundaria"]
df_educacion

anio  municipio_id municipio_nombre  promocion_efectiva_primaria  \
0     2012          6007    Adolfo Alsina                       100.26   
1     2013          6007    Adolfo Alsina                        97.35   
2     2014          6007    Adolfo Alsina                        97.36   
3     2015          6007    Adolfo Alsina                        96.00   
4     2016          6007    Adolfo Alsina                        97.67   
...    ...           ...              ...                          ...   
1480  2018          6882           Zarate                        95.42   
1481  2019          6882           Zarate                        95.62   
1482  2020          6882           Zarate                        99.51   
1483  2021          6882           Zarate                        96.77   
1484  2022          6882           Zarate                        97.49   

      promocion_efectiva_secundaria  repitencia_primaria  \
0                             84.14                 1.25   
1                             86.84                 1.68   
2                             80.79                 2.23   
3                             91.55                 2.41   
4                             95.56                 2.61   
...                             ...                  ...   
1480                          87.30                 3.83   
1481                          86.58                 3.55   
1482                          99.06                 0.04   
1483                          90.87                 2.79   
1484                          88.91                 1.35   

      repitencia_secundaria  reinscripcion_primaria  reinscripcion_secundaria  \
0                      9.06                    0.07                      1.79   
1                      5.31                    0.13                      0.12   
2                      2.52                    0.07                      0.43   
3                      5.24                    0.21                      0.85   
4                      7.27                    0.07                      1.29   
...                     ...                     ...                       ...   
1480                   8.27                    0.06                      1.80   
1481                   9.53                    0.01                      0.83   
1482                   0.00                    0.00                      0.09   
1483                   6.60                    0.17                      0.64   
1484                   7.13                    0.22                      1.11   

      abandono_interanual_primaria  ...  salidos_sin_pase_primaria  \
0                            -1.57  ...                       0.00   
1                             0.84  ...                       0.30   
2                             0.34  ...                       0.13   
3                             1.38  ...                       0.33   
4                            -0.35  ...                       0.14   
...                            ...  ...                        ...   
1480                          0.69  ...                       0.09   
1481                          0.82  ...                       0.08   
1482                          0.45  ...                       0.03   
1483                          0.28  ...                       0.03   
1484                          0.95  ...                       0.06   

      salidos_sin_pase_secundaria  salidos_con_pase_primaria  \
0                            2.71                       5.40   
1                            2.77                       5.02   
2                            2.11                       4.84   
3                            2.49                       4.96   
4                            2.79                       4.72   
...                           ...                        ...   
1480                         2.37                       2.42   
1481                         0.51                       3.30   
1482  

**(b)** Mostrar en una Serie de pandas cuantas veces entre los años 2017 y 2022 (inclusive) la repitencia es más alta en primaria que en secundaria para cada municipio.

In [141]:
mayor_repitencia_primaria_entre_2017_y_2022 = (
    df_educacion[(df_educacion["anio"] >= 2017) & (df_educacion["anio"] <= 2022)]
    .groupby("municipio_nombre")['mayor_repitencia_primaria']
    .sum()
    .sort_values(ascending=False) # los ordeno de mayor a menor
)

mayor_repitencia_primaria_entre_2017_y_2022 # Devuelve una Serie de Pandas

municipio_nombre
General Guido       5
Tordillo            4
Maipu               2
Benito Juarez       2
Pilar               2
                   ..
Coronel Suarez      0
Coronel Dorrego     0
Rauch               0
Rojas               0
Presidente Peron    0
Name: mayor_repitencia_primaria, Length: 135, dtype: int64

**(c)** Mostrar en un DataFrame la media de repitencia en primaria y la media de repitencia en secundaria agrupados por año y por zona.

In [158]:
media_repitencia_por_año_y_por_zona = (
    df_educacion
    .groupby(["anio", "zona"])[["repitencia_primaria", "repitencia_secundaria"]] # en doble corchete para mantenerlo como DataFrame
    .mean() # calcula la media para cada zona en cada año 
)

media_repitencia_por_año_y_por_zona

repitencia_primaria  repitencia_secundaria
anio zona                                                
2012 Interior             2.457350              10.699316
     Norte                1.634000               8.978000
     Oeste                2.346000               9.932000
     Sur                  2.470000              10.208750
2013 Interior             1.771368              10.499402
     Norte                1.098000               7.976000
     Oeste                1.728000               9.650000
     Sur                  1.608750              10.792500
2014 Interior             1.763333               9.676923
     Norte                1.176000               8.098000
     Oeste                1.852000               8.796000
     Sur                  1.683750               9.557500
2015 Interior             2.050427               9.586752
     Norte                1.668000               7.654000
     Oeste                2.278000               9.206000
     Sur                  1.965000               9.925000
2016 Interior             3.365556              11.697179
     Norte                2.888000              10.482000
     Oeste                3.758000              10.762000
     Sur                  3.575000              11.397500
2017 Interior             3.366581              11.529829
     Norte                2.928000               9.712000
     Oeste                3.656000              10.558000
     Sur                  3.750000              11.307500
2018 Interior             3.440513              10.145385
     Norte                2.506000               8.320000
     Oeste                3.368000               9.548000
     Sur                  3.631250              10.965000
2019 Interior             3.006667               9.070940
     Norte                2.448000               8.018000
     Oeste                3.012000               9.098000
     Sur                  3.325000               9.861250
2020 Interior             0.233761               0.094359
     Norte                0.098000               0.058000
     Oeste                0.076000               0.038000
     Sur                  0.061250               0.035000
2021 Interior             2.818632               5.453162
     Norte                1.886000               4.920000
     Oeste                3.224000               4.960000
     Sur                  2.942500               5.695000
2022 Interior             2.215983               6.536239
     Norte                1.858000               5.158000
     Oeste                2.878000               5.572000
     Sur                  3.072500               6.670000

### Regresión

In [ ]:
area_metropolitana = ['Almirante Brown', 'Avellaneda', 'Berazategui', 'Berisso', 'Brandsen', 'Campana', 'Cañuelas',
                      'Ensenada', 'Escobar', 'Esteban Echeverria', 'Exaltacion de la Cruz', 'Ezeiza',
                      'Florencio Varela', 'General Las Heras', 'General Rodriguez', 'General San Martin', 'Hurlingham',
                      'Ituzaingo', 'Jose C. Paz', 'La Matanza', 'La Plata', 'Lanus', 'Lujan', 'Lomas de Zamora',
                      'Malvinas Argentinas', 'Marcos Paz', 'Merlo', 'Moreno', 'Moron', 'Pilar', 'Presidente Peron',
                      'Quilmes', 'San Fernando', 'San Isidro', 'San Miguel', 'San Vicente', 'Tigre', 'Tres de Febrero',
                      'Vicente Lopez', 'Zarate']

### Visualización con datos geográficos

In [ ]:
# Cargamos el archivo de datos
geo_df = gpd.read_file('partidos.geojson')

In [ ]:
# Visualizamos los partidos
geo_df.plot()

In [ ]:
# Limpieza de datos. Seguir las indicaciones en la consigna del TP para este paso.


In [ ]:
# Grabamos los datos limpios en un nuevo archivo GeoJSON
geo_df.to_file('partidos_limpio.geojson', driver='GeoJSON')

In [ ]:
# Creamos un DataFrame repit con los datos de repitencia siguiendo la consigna del TP





In [ ]:
# Generamos un gráfico interactivo para el nivel de repitencia

# Si no existe chorpleth_map o alguna opcion del codigo en la siguiente celda, actualizar plotly con el siguiente comando y reiniciar el kernel.
#!pip install plotly --upgrade

# Si tienen algún error de Java después de instalar los paquetes actualizados y reiniciar el kernel, prueben refrescar el navegador con CTRL+F5.

In [ ]:
with open('partidos_limpio.geojson', 'r') as f:
    partidos = json.load(f)
    
fig = px.choropleth_map(
    repit, 
    geojson=partidos, 
    featureidkey = 'properties.cde',
    locations='cde',
    color='repitencia_secundaria',
    zoom = 4, 
    center = {"lat": -34.61315, "lon": -58.37723},
)

fig.show()

### Análisis exploratorio